### PR TITLE
MediaSession: Only Update when Playstate changes

### LINKED ISSUE #24980: Stutter related to com.google.android.tvrecommendations
### Describe the bug
Here is a clear and concise description of what the problem is:

This is a follow-up report to the forum post here: https://forum.kodi.tv/showthread.php?tid=377061 . To shortly recap, I am seeing movie playback stutter at random intervals on a Sony Bravia TV. While the stutter happens, top shows a CPU activity spike in  kworker/0:2H-mmc_complete. Disabling com.google.android.tvrecommendations seems to hide the bug.

I am streaming the content via smb from a raspberri pi from local ethernet. I believe my network setup is not to blame since Kodi works with tvrecommendations disabled. VLC plays the same content fine as well. However, I could not reproduce the problem when testing playback from an attached USB drive, although that might have been random luck.

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

The expected behavior is fairly obvious: Videos should play without stuttering.

## Actual Behavior

At random intervals (I estimate every 3-4 minutes when I am actually watching videos, and a lot less often if I try to debug the issue), the video slows down and sound stutters for a handful of seconds. 

## Possible Fix

A workaround that reliably gets rid of the problem is disabling com.google.android.tvrecommendations (aka Android TV Core Services) in the android settings.

I see AvrcpMediaPlayerWrapper activity in logcat often. Maybe disabling AVRCP responses could provide some clues what is going on.

### To Reproduce

Watch a blu-ray rip streamed via network. I also saw it happen with dvd rips, but my gut feeling is that higher bitrate content makes the bug more likely to happen. You presumably need Sony hardware with Android 10 to reproduce this. My TV's model is KD-49X8000H.

### Debuglog

The debuglog can be found here:
I have attached a kodi log and adb logcat in logs.zip. The stutter happens at 2024-04-13 00:08:42 ++ . Note that in this particular case I changed the volume around the time it happened, but the bug also happens without any overlays showing up or pressing any buttons. I can try to catch the bug again without any interactive input if requested.

### Screenshots 
Here are some links or screenshots to help explain the problem:

There are no problems visible in a screenshot. I can record the screen of the TV with a camera if necessary.

### Your Environment
Used Operating system:

 - [X] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: 10
 - Kodi version: 20.5 (from google play store).

I also tried to downgrade kodi down to 19.0 because I believe the issue is relatively new. Going back to 19.0 did *not* help, making it likely that a system update introduced the issue.

I tried to wipe all kodi data, without success. I performed a factory reset of the entire Android system and disabled automatic Google Play Store updates, also without any impact on the problem. The last firmware update by Sony was distributed about a year ago I believe.

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by fritsch**
This reduces the load of the builder, especially it makes limited sense to send this information so often. Reason for this is that the MediaSession has the speed, the currenttime and even when a seek with SetPosition happens it therefore knows the state.

This optimizes the interaction with the MediaSession and solves real life problem -> Stuttering on low power boxes.

## Description
On every playback kodi creates a MediaSession and updates this twice per second during playback. This is run off the mainthread can cause stalling video when running its builders. Furthermore Android OS in the background fires a vaste amount of AVRCP Status Updates are not needed, cause the receiver can calculate the proper fields as long as no state changes he does not know off.

## Motivation and context
The AVRCP logspam is a direct impact from updating PlaybackStatus twice a second.

## How has this been tested?
It was not tested - I just read the code.

## What is the effect on users?
When playing a Video the MediaSession / Leanback information is only updated when the state changes. All other information can be computed from the provided information.

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

Honestly speaking I have no idea what the original did and why such a high frequency was needed - I wondered basically forever why I get so much AVRCP logs in my adb shell while I have zero to do with Bluetooth there.

See also: https://github.com/xbmc/xbmc/issues/24980 for a discussion.

### CODE REVIEW

File: xbmc/platform/android/activity/XBMCApp.h [hunk 1]
```diff
@@ -258,6 +258,7 @@ class CXBMCApp : public IActivityHandler,
   bool m_wakeUp{false};
   bool m_aeReset{false};
   bool m_hdmiPlugged{true};
+  unsigned int m_MediaSessionVideoUpdateCounter = 0;
```
Comment (ksooo) [inline xbmc/platform/android/activity/XBMCApp.h:261]: We usually start member variable names lower case after the m_ prefix.

**Comment by fritsch** [comment]
Example:

04-13 00:08:48.673  3288  3288 V AvrcpMediaPlayerWrapper: onPlaybackStateChanged(): org.xbmc.kodi : PlaybackState {state=3, position=89011, buffered position=0, speed=1.0, updated=358864, actions=-1, custom actions=[], active item id=-1, error=null}
04-13 00:08:48.674  6094  6094 D BasicMediaStatusFactory: Supported media session actions: -1
04-13 00:08:48.674  3288  3288 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
04-13 00:08:48.676  3288  3288 V AvrcpMediaPlayerWrapper: trySendMediaUpdate(): Metadata has been updated for org.xbmc.kodi
04-13 00:08:48.677  3288  3288 D AvrcpMediaPlayerList: sendMediaUpdate
04-13 00:08:48.677  3288  3288 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 00:08:48.677  3288  3288 D AvrcpTargetService: onMediaUpdated: track_changed=false state=false queue=false
04-13 00:08:48.677  3288  3288 D AvrcpNativeInterface: sendMediaUpdate: metadata=false playStatus=false queue=false
04-13 00:08:48.677  3288  3288 D AvrcpTargetJni: sendMediaUpdateNative
04-13 00:08:48.677  3288  3288 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 00:08:48.685  2345  2345 I hwcomposer: same layer type is change to 3 from 3 for the same layer id
04-13 00:08:48.967  2345  2345 I chatty  : uid=1000(system) composer@2.1-se identical 17 lines
04-13 00:08:48.984  2345  2345 I hwcomposer: same layer type is change to 3 from 3 for the same layer id
04-13 00:08:48.986  2342  7184 D mtdrm_WV: MTAL: _oecc44 OEMCrypto_GetHDCPCapability entry 
04-13 00:08:48.986  2342  7184 D mtdrm_WV: MTAL: _oecc44 OEMCrypto_GetHDCPCapability exist current=255,maximum=255
04-13 00:08:49.017  2345  2345 I hwcomposer: same layer type is change to 3 from 3 for the same layer id
04-13 00:08:49.167  2345  2345 I chatty  : uid=1000(system) composer@2.1-se identical 7 lines
04-13 00:08:49.184  2345  2345 I hwcomposer: same layer type is change to 3 from 3 for the same layer id
04-13 00:08:49.186  3288  3288 V AvrcpMediaPlayerWrapper: onPlaybackStateChanged(): org.xbmc.kodi : PlaybackState {state=3, position=89656, buffered position=0, speed=1.0, updated=359380, actions=-1, custom actions=[], active item id=-1, error=null}
04-13 00:08:49.187  6094  6094 D BasicMediaStatusFactory: Supported media session actions: -1
04-13 00:08:49.187  3288  3288 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
04-13 00:08:49.190  3288  3288 V AvrcpMediaPlayerWrapper: trySendMediaUpdate(): Metadata has been updated for org.xbmc.kodi
04-13 00:08:49.190  3288  3288 D AvrcpMediaPlayerList: sendMediaUpdate
04-13 00:08:49.190  3288  3288 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 00:08:49.190  3288  3288 D AvrcpTargetService: onMediaUpdated: track_changed=false state=false queue=false
04-13 00:08:49.190  3288  3288 D AvrcpNativeInterface: sendMediaUpdate: metadata=false playStatus=false queue=false
04-13 00:08:49.190  3288  3288 D AvrcpTargetJni: sendMediaUpdateNative

File: xbmc/platform/android/activity/XBMCApp.h (hunk 1 above)
Comment (fritsch) [reply xbmc/platform/android/activity/XBMCApp.h:261]: I will fix that immediately. Actually I am sitting in the sun ... and was happy that I produced something that looked like kodi code - with just 300 NITs.

**Comment by ksooo** [comment]
> makes limited sense to send video information every half a second, because position unknown is not really adding any value. A video normally runs for some hours, therefore one update per minute seems okayish to me.

This holds true for „classical“ movies and tv shows, but there is no guarantee that all video duels are of that kind. TV channels may switch video resolution etc while annoying us with ads at any point in time, to give an example.

**Comment by fritsch** [comment]
> > makes limited sense to send video information every half a second, because position unknown is not really adding any value. A video normally runs for some hours, therefore one update per minute seems okayish to me.
> 
> This holds true for „classical“ movies and tv shows, but there is no guarantee that all video streams are of that kind. TV channels may switch video resolution etc while annoying us with ads at any point in time, to give an example.

All good. Where do you see a real problem?

- Video  is less than 60 seconds: Stats are updated on the start and on stop
- Video is less than 120 seconds: Stats are updated on start, once more and on stop

**Comment by fritsch** [comment]
@ksooo Asking differently. Do you know what happens with the information we send out of kodi? Who is the user of those stats?

**Comment by fritsch** [comment]
From my FireTV device:

04-13 17:31:09.634 20244 20244 E BluetoothAvrcpController: Could not bind to Bluetooth Service with Intent { act=android.bluetooth.IBluetoothAvrcpController }
04-13 17:31:09.723 21261 22311 W ActivityManager: Unable to start service Intent { act=android.media.browse.MediaBrowserService cmp=com.android.bluetooth/.avrcpcontroller.BluetoothMediaBrowserService } U=10: not found
04-13 17:31:11.215 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
04-13 17:31:11.219 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:31:11.223 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:32:13.214 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
04-13 17:32:13.216 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:32:13.220 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:33:15.489 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
04-13 17:33:15.492 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:33:15.493 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:34:17.507 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
04-13 17:34:17.510 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:34:17.510 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:35:19.175 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
04-13 17:35:19.178 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue

after the change - it seems our Update causes something we did not intend at all ... namely that bluetooth wrapper player of the Android OS going nuts ... yeah ... let's keep that one open here, maybe someone comes by who knows what we are doing there :-)

**Comment by fritsch** [comment]
Okay I read through the docu: https://developer.android.com/reference/android/media/session/MediaSession - and also tested on my FireTV with youtube app:

04-13 17:51:16.449 21634 21634 I AvrcpMediaPlayerList: Adding wrapped media player: com.amazon.firetv.youtube at key: 3
04-13 17:51:32.791 21634 21634 I AvrcpMediaPlayerList: onMediaKeyEventSessionChanged: token=com.amazon.firetv.youtube
04-13 17:51:32.800 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:51:32.806 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=1 :  play_state=1 :  queue=1
04-13 17:51:32.810 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:51:32.811 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:51:32.966 21634 21634 W AvrcpMediaPlayerWrapper: onPlaybackStateChanged(): com.amazon.firetv.youtube tried to update with no new data
04-13 17:51:32.967 21634 21634 W AvrcpMediaPlayerWrapper: onPlaybackStateChanged(): com.amazon.firetv.youtube tried to update with no new data
04-13 17:51:32.982 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:51:32.982 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:51:34.575 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:51:34.575 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=1 :  queue=0
04-13 17:51:34.580 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:51:34.580 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=1 :  play_state=0 :  queue=1
04-13 17:51:34.778 21634 21634 E AvrcpMediaPlayerWrapper: The callback metadata doesn't match controller metadata
04-13 17:51:34.779 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:51:34.779 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:51:34.781 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:51:34.781 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:52:57.567 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:52:57.567 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=1 :  queue=0
04-13 17:52:57.570 21634 21634 W AvrcpMediaPlayerWrapper: onPlaybackStateChanged(): com.amazon.firetv.youtube tried to update with no new data
04-13 17:52:57.575 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:52:57.575 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:53:03.990 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:53:03.991 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=1 :  queue=0
04-13 17:53:04.019 21634 21634 W AvrcpMediaPlayerWrapper: onPlaybackStateChanged(): com.amazon.firetv.youtube tried to update with no new data
04-13 17:53:04.020 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:53:04.020 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0
04-13 17:53:04.479 21634 21634 W AvrcpMediaPlayerWrapper: onPlaybackStateChanged(): com.amazon.firetv.youtube tried to update with no new data
04-13 17:53:04.482 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:53:04.482 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=0 :  play_state=0 :  queue=0

Here basically "not much" happens - though there is a difference:
 04-13 17:32:13.214 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state

Is not matching and most likely causes this "Error" to be logged. Basically that could mean, we register us wrongly

Kodi:

04-13 17:55:46.481 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:55:46.485 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=1 :  play_state=1 :  queue=1
04-13 17:56:28.993 21634 21634 W AvrcpMediaPlayerList: mediaUpdatedCallback(): Tried to update with null state
04-13 17:56:29.003 21634 21634 I AvrcpMediaPlayerList: sendMediaUpdate: Creating a one item queue for a player with no queue
04-13 17:56:29.003 21634 21634 I bt_stack: [INFO:avrcp_service.cc(346)] virtual void bluetooth::avrcp::AvrcpService::SendMediaUpdate(bool, bool, bool) track_changed=1 :  play_state=1 :  queue=1

**Comment by fritsch** [comment]
For now it's not even reviewed. The reason I added backport is, that we currently hammer twice per second on devices - and have no idea why we are doing that.

Let's just keep it open for discussion.

**Comment by ksooo** [comment]
I'm definitely open for discussions re. a backport. However, it should have priority to avoid regressions for Omega. And in case we do not know what these data are good for I have, well, not the best feeling about risk. We're hammering at those deices for like ages? And now this becomes important? Why? Because one user complained? Not really.

**Comment by fritsch** [comment]
Actually I simply set the "Backport Needed" too early, that's all. I fully agree with your arguments given. Let me put it that way: If we can solve and understand it properly - it might become a candidate. That's it.

**Comment by sundermann** [comment]
If this causes too much stress on the main thread wouldn't it make sense to do it inside a thread?

**Comment by fritsch** [comment]
Will need: https://github.com/xbmc/libandroidjni/pull/49 so that we can properly register the Actions and not using a 0xffffffffff mask, which even enables what we don't support.

**Comment by fritsch** [comment]
> If this causes too much stress on the main thread wouldn't it make sense to do it inside a thread?

Yeah - but - not so easy on android where the activity needs to do it?

**Comment by sundermann** [comment]
Was just curious ;)

**Comment by ksooo** [comment]
> Actually I simply set the "Backport Needed" too early, that's all. I fully agree with your arguments given. Let me put it that way: If we can solve and understand it properly - it might become a candidate. That's it.

Absolutely. Thanks.

**Comment by a1rwulf** [comment]
> For now it's not even reviewed. The reason I added backport is, that we currently hammer twice per second on devices - and have no idea why we are doing that.
> 
> Let's just keep it open for discussion.

If remote controls use the media session to show progress bars it makes perfect sense to update twice a second, we poll kodi’s player state every 300ms in order to not run out of sync with the actual playstate. That might be the reason.

**Comment by fritsch** [comment]
> > For now it's not even reviewed. The reason I added backport is, that we currently hammer twice per second on devices - and have no idea why we are doing that.
> > Let's just keep it open for discussion.
> 
> If remote controls use the media session to show progress bars it makes perfect sense to update twice a second, we poll kodi’s player state every 300ms from out remote application in order to not run out of sync with the actual playstate. That might be the reason.

That's what I said in the title and which is why I kept it for audio only. For a movie I don't think that buffer updates twice per second have any real meaning for a user looking at the numbers, while this is a (silly or not) workaround, the error below is most likely the root-cause we write on EMMCs as some TVs write Errors to disk.

Checking with youtube (exxoplayer) the updates seem only to happen on startup or they send it in a certain way that the receiver does not error out.

Fixing this:
 04-13 17:32:13.214 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
 ...
 04-13 17:56:28.993 21634 21634 W AvrcpMediaPlayerList: mediaUpdatedCallback(): Tried to update with null state

might maybe solve the issue. But I currently don't see where we send a "wrong state".

**Comment by fritsch** [comment]
Reason for youtube is this:
https://stackoverflow.com/questions/48444213/mediacontrollercompat-get-state-doesnt-return-the-current-position

"You don't need to update the PlaybackState every second. The MediaSession remembers the time at which you call setPlaybackState() so when you call getPosition() and the state is playing, it will automatically add the difference between the current time and the last update time to the position (also taking playback speed into account)"

^^ I think it makes therefore no sense at all to send any new PlaybackState Message if we did not change anything. Means: As long as we stay playing we should not send anything at all.

**Comment by fritsch** [comment]
I will update this PR with a minimal fixup to keep the chance for a backport. I don't have such a remote / setup to test sadly, therefore would be happy if someone can test the change if it works properly or if we have to update more often.

**Comment by stefand** [comment]
Here is a semi-related question, pardon me if this doesn't belong here: Might Kodi set a wrong state when playback is paused? I think it might tell Android that playback is stopped entirely, which might not be the right thing.

An observation I made is that when switching out of Kodi (home button on the remote, not the exit button inside kodi) the playback notication disappears within fractions of a second. When going back to kodi, the notification shows up, although it is moot within fractions of a second again because the home screen is replaced by Kodi.

Spotify leaves the played song around, but it also keeps producing audio while out of focus.

My inner Wine dev says "maybe, but in any case this belongs in a different MR".

**Review by thexai — APPROVED** [review]
I completely agree with the reasoning. In fact I was have same symptom mentioned in the issue on Shield:

> While the stutter happens, top shows a CPU activity spike in kworker/0:2H-mmc_complete. Disabling com.google.android.tvrecommendations seems to hide the bug.

In my case it was only playing Dolby Vision files and I thought they were cache problems, etc. It was totally random: it could happen after 20 minutes, 1 hour or not happen at all in an entire movie.

The important thing is that the solution was the same: disable `com.google.android.tvrecommendations` and some others processes (I don't remember exactly).

As playing Dolby Vision already uses more CPU (not specifically in Kodi but other internal Shield processes) that's why I think that in my case it only happened with DV.

I even found a relationship with temperature. At a higher temperature and longer operating time this happened sooner. It is logical because from a certain temperature the CPU is limited (throttling).

I even changed the thermal paste on the Shield CPU.... but it is true that since I did these 2 things it has never happened once again. 

If not remember wrong, from this there have been no more Shield firmware updates, so I still have these processes disabled.

**Comment by thexai** [comment]
Other users sporadically report this same issue:

https://www.reddit.com/r/kodi/comments/12mdkd1/4k_dolby_vision_occasional_stutter_maven/

**Comment by fritsch** [comment]
I added the formating changes. Key question to me still is: Does the original functionality someone might use still work? Don't want to cause regressions for people relaying on this feature, as I still have no idea how they do it.

**Comment by a1rwulf** [comment]
> I added the formating changes. Key question to me still is: Does the original functionality someone might use still work? Don't want to cause regressions for people relaying on this feature, as I still have no idea how they do it.

We are in alpha now, I'd say merge and see if anybody complains.

**Review by a1rwulf — APPROVED** [review]

File: xbmc/platform/android/activity/XBMCApp.cpp [hunk 2]
```diff
@@ -907,9 +917,13 @@ void CXBMCApp::UpdateSessionState()
   else
     state = CJNIPlaybackState::STATE_STOPPED;
 
-  builder.setState(state, pos, speed, CJNISystemClock::elapsedRealtime())
-      .setActions(CJNIPlaybackState::PLAYBACK_POSITION_UNKNOWN);
-  m_mediaSession->updatePlaybackState(builder.build());
+  if ((oldPlayState != m_playback_state) || !m_mediaSessionUpdated)
+  {
+    builder.setState(state, pos, speed, CJNISystemClock::elapsedRealtime())
+        .setActions(CJNIPlaybackState::PLAYBACK_POSITION_UNKNOWN);
```
Comment (joseluismarti) [inline xbmc/platform/android/activity/XBMCApp.cpp:923]: I've not checked if it affects and could be corrected later, `setActions()` should receive as a parameter the current actions available in this session, not a position value.

https://developer.android.com/reference/android/media/session/PlaybackState.Builder#setActions(long)
https://developer.android.com/training/tv/playback/media-session#state

Comment (joseluismarti) [reply xbmc/platform/android/activity/XBMCApp.cpp:923]: Sorry, I didn't see your [previous comment](https://github.com/xbmc/xbmc/pull/24986#issuecomment-2053701102)

Comment (fritsch) [reply xbmc/platform/android/activity/XBMCApp.cpp:923]: I would keep the MASK for now as it is - 0xFFFFFFFFFF would enable them all, which is not correct - but without libjniandroid changes I cannot change that. I would do that afterwards to keep this one out of the other one.

I would need your help here:

> 04-13 17:32:13.214 21634 21634 E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state
> 04-13 17:56:28.993 21634 21634 W AvrcpMediaPlayerList: mediaUpdatedCallback(): Tried to update with null state

Especially the "Error" is something youtube does not produce and due to my limiting setup capabilities I don't have an idea how to debug that.

Comment (fritsch) [reply xbmc/platform/android/activity/XBMCApp.cpp:923]: Corresponding libjniandroid change: https://github.com/xbmc/libandroidjni/pull/49 - would be great if could give your okay as well.

Comment (joseluismarti) [reply xbmc/platform/android/activity/XBMCApp.cpp:923]: libjniandroid change is ok, all these constants are available at API level 24. 

I'll try to find out what is causing this error message

**Comment by fritsch** [comment]
Will backport after I got some feedback in the forum where I posted the testbuilds.

File: xbmc/platform/android/activity/XBMCApp.cpp (hunk 2 above)
Comment (joseluismarti) [reply xbmc/platform/android/activity/XBMCApp.cpp:923]: >  E AvrcpMediaPlayerWrapper: The callback playback state doesn't match the current state

In theory this is not an error, when `builder.setState(...)` is called, if more than 500 ms have passed, a different state is considered.

Look at these two methods:
https://android.googlesource.com/platform/packages/apps/Bluetooth/+/master/src/com/android/bluetooth/audio_util/MediaPlayerWrapper.java#468
https://android.googlesource.com/platform/packages/apps/Bluetooth/+/master/src/com/android/bluetooth/audio_util/MediaPlayerWrapper.java#529

> W AvrcpMediaPlayerList: mediaUpdatedCallback(): Tried to update with null state

I don't see this message in the logcat

Comment (fritsch) [reply xbmc/platform/android/activity/XBMCApp.cpp:923]: Okay, then we keep it as is, explains at least why I did not find anything obviously wrong in the implementation concerning this part.

The null state I saw on my FireTV, while we definitely always send a proper state - so then let's ignore this for now.

**Comment by matejdro** [comment]
I think this might have caused a regression: https://github.com/xbmc/xbmc/issues/25838

**Comment by CastagnaIT** [comment]
@matejdro you opened an issue for kodi 21, instead this PR is on Kodi 22 only

**Comment by matejdro** [comment]
It looks like it was backported to 21.1: https://github.com/xbmc/xbmc/pull/25041

**Comment by CastagnaIT** [comment]
oh right was forgotten to add flags

**Comment by fritsch** [comment]
Thanks for finding this. That's the first outtage ;-) good catch, let's fix it.

**Pull request merged**